users/vieting/experiments/switchboard/hybrid/feat/specaug_sorted.py

```diff
@@ -0,0 +1,173 @@
+    # https://timvieira.github.io/blog/post/2014/08/01/gumbel-max-trick-and-weighted-reservoir-sampling/
+    z = -tf.log(-tf.log(tf.random_uniform((n_batch, tf.shape(x)[axis]), 0, 1)))
+    _, indices = tf.nn.top_k(z, num if isinstance(num, int) else tf.reduce_max(num))
+    # indices should be sorted, and of shape (batch,num), entries (int32) in [0,dim)
+    if isinstance(num, int):
+        for i in range(num):
+            x = _mask(
+                x,
+                batch_axis=batch_axis,
+                axis=axis,
+                pos=indices[:, i],
+                max_amount=max_dims,
+                sorted_indices=sorted_indices,
+            )
+    else:
+        _, x = tf.while_loop(
+            cond=lambda i, _: tf.less(i, tf.reduce_max(num)),
+            body=lambda i, x: (
+                i + 1,
+                tf.where(
+                    tf.less(i, num),
+                    _mask(
+                        x,
+                        batch_axis=batch_axis,
+                        axis=axis,
+                        pos=indices[:, i],
+                        max_amount=max_dims,
+                        sorted_indices=sorted_indices,
+                    ),
+                    x,
+                ),
+            ),
+            loop_vars=(0, x),
+        )
+    return x
+
+
+def specaugment_eval_func(data, network, mask_size, time_factor=1):
+    x = data.placeholder
+    from returnn.tf.compat import v1 as tf
+
+    filter_layer = network.layers["conv_h_filter"].output.placeholder
+    sorted_idce = sort_filters_by_center_freq(filter_layer)
+    # summary("features", x)
+    step = network.global_train_step
+    step1 = tf.where(tf.greater_equal(step, 1000), 1, 0)
+    step2 = tf.where(tf.greater_equal(step, 2000), 1, 0)
+
+    def get_masked():
+
+        x_masked = _random_mask(
+            x,
+            batch_axis=data.batch_dim_axis,
+            axis=data.time_dim_axis,
+            min_num=step1 + step2,
+            max_num=tf.maximum(tf.shape(x)[data.time_dim_axis] // 100, 2) * (1 + step1 + step2 * 2),
+            max_dims=20 // time_factor,
+        )
+
+        # Apply freq masking
+        x_masked = _random_mask(
+            x_masked,
+            batch_axis=data.batch_dim_axis,
+            axis=data.feature_dim_axis,
+            min_num=step1 + step2,
+            max_num=2 + step1 + step2 * 2,
+            max_dims=data.dim // mask_size,
+            sorted_indices=sorted_idce,
+        )
+        return x_masked
+
+    x = network.cond_on_train(get_masked, lambda: x)
+    return x
+
+
+def specaug_layer_sorted(in_layer, mask_size):
+    """
+    specaug layer with default hybrid settings
+
+    :param in_layer:
+    """
+    return {
+        "class": "eval",
+        "from": in_layer,
+        "eval": "self.network.get_config().typed_value('specaugment_eval_func')("
+        "source(0, as_data=True, auto_convert=False), "
+        "network=self.network),"
+        " "+str(mask_size),
```

Comment:
Are you sure this works? Have you tested it? It looks like it's going to result in syntax errors. Please always test things before creating a PR.